keps/0009-operator-toolkit.md

```diff
@@ -0,0 +1,287 @@
+
+#### Files
+
+If a filename is specified, KUDO will execute a Go Template on the relevant filename, described more in detail in [Templates](#templates).
+
+#### Webhooks
+
+If a webhook is specified, KUDO will POST the listed webhook with a JSON body containing the full KUDO [execution state](#execution-state). The webhook **MUST** respond with the following:
+
+- Status code: 200
+- Content-type: application/json
+
+The body of this response **MUST** be a fully-qualified Kubernetes object in JSON form. If multiple objects are required, they **MUST** be wrapped in a Kubernetes List API object. The contents of a list object **CAN** be composed of multiple Kubernetes objects of different types.
+
+Webhooks **CAN** have side effects to control external state or to fetch external parameters for use in their templates, as long as the response is still a Kubernetes object.
+
+Webhook HTTP connections time out after 30 seconds, after which a Timeout event is added to the PlanExecution the task is running on.
+
+#### Resources vs. Patches
+
+KUDO additionally supports Kustomize for defining resources and patching for tasks. Kustomize is applied after any Go templating steps. This is useful for defining common bases for objects, or for extending frameworks as described in [Extensions and Bases](#extensions-and-bases).
+
+#### Task Application
+
+Once KUDO has assembled a full set of templates for a task, they will be applied using Kubernetes server-side apply at the same time.
+
+### Parameters
+
+Parameters are a key-value list of parameter names and their defaults. As described in [Execution State](#execution-state), they are wrapped into the `.Params` object for use by all templated objects.
+
+### Templates
+
+A template is a standard Kubernetes manifest which **MAY** have additional Go Templating. These Go Templates include [Sprig](https://github.com/masterminds/sprig) to provide a standard library of functions familiar to Helm users. Additional values are available and are described in [Execution State](#execution-state).
+
+## Extensions and Bases
+
+Extensions and bases describe a mechanism for building operators or extensions to operators from a given base. As an example, a base can be an operator, a Helm chart, a CNAB bundle, or any future format that describes the deployment of a set of resources.
+
+In this document an **extension** is any KUDO operator that extends from some base. A **base** is the complete set of manifests, metadata, and other files provided by that base's type. A base should provide the complete information that users of that base tool are expected to have. The base types and what they expose to charts that extend from them are described in their respective sub-sections.
+
+To support extending from a base, `operator.yaml` is extended to support the `extends` keyword:
+
+```yaml
+extends:
+  kudo:
+    operator: "mysql"
```

Comment:
maybe this should be `framework`? Unless we're renaming all our Framework and FrameworkVersions

Reply:
idk @ryadav88 is for calling them "operator", I'm ok either way. KUDO operator is a little redundant in the ATM machine sense.

Reply:
IMO, we should standardize on `operator` both here as well as in `kudobuilder/frameworks` repo. 

"We have an Operator builder, KUDO using which you can develop `frameworks`. Next Q: What is a framework?  - you heard of Operators. That's what it is. "

Reply:
+1 on calling it operator and retiring framework. The less terminology we have to explain, the better.